Bug/fix bug on api

radhakrishnan-getfudogroup opened the pull request:
Hi @mzrghorbani sir  ,

I have made the following updates in this Pull Request:
**1.Added a new API in the Device module — Fetch Devices by Site ID.**
**Request URl:**
`http://{BaseURL}/api/v1/devices/devices/{Site_id)
`
**Response Body(sample):**

[```

  {
    "site_id": 4,
    "device_id": "pos-terminal-129",
    "name": "POS Terminal 1",
    "ip_address": "192.168.1.10"
  },
  {
    "site_id": 4,
    "device_id": "pos-terminal-110",
    "name": "POS Terminal 1",
    "ip_address": "192.168.1.10"
  }
]
```
**2. Updated column names in the UserRegister endpoint** 

Please review and let me know if any changes are required. Thank you

mzrghorbani reviewed (changes requested):
PR #42 Review - Critical Issues to Fix
### Problem Summary
The endpoint confuses business IDs (strings) with database IDs (integers).

### Database Schema:
```
Site:    id (int) = 1, 2, 3...  |  site_id (string) = "site-123"
Device:  id (int) = 1, 2, 3...  |  device_id (string) = "pos-001"
         site_id (int FK) = points to sites.id
```
Rule: APIs must ONLY use business IDs (strings), never database IDs (integers).

### Required Fixes
1. Fix [DevicesEndpoints.py](vscode-file://vscode-app/c:/Users/mazia/AppData/Local/Programs/Microsoft%20VS%20Code/resources/app/out/vs/code/electron-browser/workbench/workbench.html) (Line 170)
Current (Risky):

```
@router.get("/devices/{site_id}", tags=["Devices"])
async def get_device_site_id(site_id: int) -> List[Dict[str, Any]]:  # ❌ int
    device = await db_service.get_device_by_site_id(site_id)
    return [{
        "site_id": d.site_id,  # ❌ Returns integer FK
        "id": d.id,            # ❌ Never expose internal IDs
        ...
    }]
```
Correct:

```
@router.get("/sites/{site_id}/devices", tags=["Devices"])
async def get_devices_by_site(site_id: str) -> List[Dict[str, Any]]:  # string
    db_service = await get_database_service()
    
    # Verify site exists
    site = await db_service.get_site_by_site_id(site_id)
    if not site:
        raise HTTPException(status_code=404, detail=f"Site '{site_id}' not found")
    
    # Get devices
    devices = await db_service.get_devices_by_site_id(site_id)
    
    return [{
        "site_id": site.site_id,      # String: "site-123"
        "device_id": d.device_id,      # String: "pos-001"
        "name": d.name,
        "device_type": d.device_type,
        "status": d.status,
        "ip_address": d.ip_address,
    } for d in devices]
```

2. Add Method to [database.py](vscode-file://vscode-app/c:/Users/mazia/AppData/Local/Programs/Microsoft%20VS%20Code/resources/app/out/vs/code/electron-browser/workbench/workbench.html)

```
async def get_devices_by_site_id(self, site_id: str) -> List[Device]:
    """Get devices by site_id string (e.g., 'site-123')."""
    from sqlalchemy import select
    
    async with self.get_session() as session:
        # Get site first
        site = await self.get_site_by_site_id(site_id)
        if not site:
            return []
        
        # Get devices using internal site.id
        result = await session.execute(
            select(Device).where(
                Device.site_id == site.id,  # Use internal ID here
                Device.is_active.is_(True)
            ).order_by(Device.created_at.desc())
        )
        return list(result.scalars().all())
```

3. Rename Existing Method (Line 183)

```
# Old confusing name
async def get_device_by_site_id(self, site_id: int)

# New clear name
async def get_devices_by_site_internal_id(self, site_internal_id: int)
```
Key Rules
1. API parameters: Always [site_id: str](vscode-file://vscode-app/c:/Users/mazia/AppData/Local/Programs/Microsoft%20VS%20Code/resources/app/out/vs/code/electron-browser/workbench/workbench.html), [device_id: str](vscode-file://vscode-app/c:/Users/mazia/AppData/Local/Programs/Microsoft%20VS%20Code/resources/app/out/vs/code/electron-browser/workbench/workbench.html)
2. API responses: Only return business IDs (strings)
3. Never expose: [id](vscode-file://vscode-app/c:/Users/mazia/AppData/Local/Programs/Microsoft%20VS%20Code/resources/app/out/vs/code/electron-browser/workbench/workbench.html), [site.id](vscode-file://vscode-app/c:/Users/mazia/AppData/Local/Programs/Microsoft%20VS%20Code/resources/app/out/vs/code/electron-browser/workbench/workbench.html), [device.id](vscode-file://vscode-app/c:/Users/mazia/AppData/Local/Programs/Microsoft%20VS%20Code/resources/app/out/vs/code/electron-browser/workbench/workbench.html) fields
4. URL pattern: [/sites/{site_id}/devices](vscode-file://vscode-app/c:/Users/mazia/AppData/Local/Programs/Microsoft%20VS%20Code/resources/app/out/vs/code/electron-browser/workbench/workbench.html) not [/devices/{site_id}](vscode-file://vscode-app/c:/Users/mazia/AppData/Local/Programs/Microsoft%20VS%20Code/resources/app/out/vs/code/electron-browser/workbench/workbench.html)

You can test this by:

```
# Should work with STRING site_id
curl http://localhost:8000/api/v1/sites/site-123/devices

# Expected response (all strings):
[{
  "site_id": "site-123",
  "device_id": "pos-terminal-001",
  "name": "POS Terminal 1",
  ...
}]
```

I hope this helps, but please let me know if it is still unclear.

radhakrishnan-getfudogroup commented:
Hi, @mzrghorbani  sir,
Thanks for the Review.
I've completed the requested changes and also added support for the new GET /devices?site_id= endpoint required by the frontend integration .

**Changes Made**
**1. Replaced all site.id references with site.site_id**
For consistency and to follow the recommended security practice (avoiding direct integer primary key exposure), all queries and relationships now use the site_id string identifier instead of the internal numeric id.
Updated files:
```
backend/src/homepot/database.py
backend/src/homepot/main.py
backend/src/homepot/models.py
backend/tests/test_database.py
backend/tests/test_models.py
```

**2. Updated Device model foreign key**
Originally Device.site_id referenced the integer Site.id.
Since we must not expose/use the numeric ID externally, and the frontend needs to query by site_id, I changed:

`the site_id column on the Device table from int → str
`

`the foreign key to reference Site.site_id instead of Site.id
`
This aligns the schema with the business rules and ensures Device records can be fetched reliably using the external site_id.
**3. Added GET API to fetch devices by site_id**
Implemented the endpoint required by the frontend, ensuring it uses site_id (string) instead of internal DB IDs.
Request for Review
Please review this workflow and confirm whether the model, database, and API changes align with project standards.
If everything looks correct, I’ll proceed with any final cleanup or migration adjustments.

Thanks.

mzrghorbani reviewed (approved):
Ready for merge.
